packages/hw-transport-mocker/src/RecordStore.js

```diff
@@ -1,41 +1,108 @@
 //@flow
 
-export default class RecordStore {
-  cache: { [_: string]: string } = {};
+export function RecordStoreInvalidSynthax(message: string) {
+  this.name = "RecordStoreInvalidSynthax";
+  this.message = message;
+  this.stack = new Error().stack;
+}
+//$FlowFixMe
+RecordStoreInvalidSynthax.prototype = new Error();
+
+export function RecordStoreQueueEmpty() {
+  this.name = "RecordStoreQueueEmpty";
+  this.message = "EOF: no more APDU to replay";
+  this.stack = new Error().stack;
+}
+//$FlowFixMe
+RecordStoreQueueEmpty.prototype = new Error();
+
+export function RecordStoreWrongAPDU(expected: string, got: string) {
+  this.name = "RecordStoreWrongAPDU";
+  this.message = `wrong apdu to replay. Expected ${expected}, Got ${got}`;
+  this.expectedAPDU = expected;
+  this.gotAPDU = got;
+  this.stack = new Error().stack;
+}
+//$FlowFixMe
+RecordStoreWrongAPDU.prototype = new Error();
+
+export function RecordStoreRemainingAPDU(expected: string) {
+  this.name = "RecordStoreRemainingAPDU";
+  this.message = `replay expected more APDUs to come:\n${expected}`;
+  this.stack = new Error().stack;
+}
+//$FlowFixMe
+RecordStoreRemainingAPDU.prototype = new Error();
+
+export type Queue = [string, string][];
+
+export class RecordStore {
+  queue: Queue;
+
+  constructor(queue?: ?Queue) {
+    this.queue = queue || [];
+  }
+
+  isEmpty = () => this.queue.length === 0;
 
   recordExchange(apdu: Buffer, out: Buffer) {
-    const { cache } = this;
+    this.queue.push([apdu.toString("hex"), out.toString("hex")]);
+  }
+
+  replayExchange(apdu: Buffer): Buffer {
+    const { queue } = this;
+    if (queue.length === 0) {
+      throw new RecordStoreQueueEmpty();
+    }
+    const [head, ...rest] = queue;
```

Comment:
what about renaming `...rest` to `...tail`? :p

Reply:
done!